### PR TITLE
Enable local caching in byte_stream_server_proxy

### PR DESCRIPTION
**Pull request by iain-macdonald**
This PR contains the changes in https://github.com/buildbuddy-io/buildbuddy/pull/7048 -- review that one first.

This PR makes `byte_stream_server_proxy` use the `byte_stream_server` installed in the environment by `cache_proxy` in https://github.com/buildbuddy-io/buildbuddy/pull/7048 for caching data locally. It attempts to serve reads out of the local bytestreamserver, when the data is available, and writes data that's read through the proxy to the local bytestreamserver. There is a bit of a hack that converts downloaded resource names into uploaded resource names, but otherwise it's mostly just manipulating gRPC streams.

Note that the test is largely copied from `server/remote_cache/byte_stream_server/byte_stream_server_test.go` because that test has pretty good coverage of compression cases that I wanted to keep. I factored out some common code into a testutil.

**Related issues**: N/A

### CODE REVIEW

**Review by tylerwilliams — COMMENTED** [review]
The stream code is so hard to follow (not your fault, grpc streams are tough). My understanding of this code is that it's starting async jobs during the write to populate the local cache. I wonder if a simpler approach where we just copy files from remote->local once a write succeeds would work?

Alternatively, we could do the copy synchronously, chunk by chunk, and if any error is encountered locally just set a flag and skip subsequent local writes?

wdyt?

**Comment by iain-macdonald** [comment]
> The stream code is so hard to follow (not your fault, grpc streams are tough). My understanding of this code is that it's starting async jobs during the write to populate the local cache. I wonder if a simpler approach where we just copy files from remote->local once a write succeeds would work?
> 
> Alternatively, we could do the copy synchronously, chunk by chunk, and if any error is encountered locally just set a flag and skip subsequent local writes?
> 
> wdyt?



> The stream code is so hard to follow (not your fault, grpc streams are tough). My understanding of this code is that it's starting async jobs during the write to populate the local cache. I wonder if a simpler approach where we just copy files from remote->local once a write succeeds would work?
> 
> Alternatively, we could do the copy synchronously, chunk by chunk, and if any error is encountered locally just set a flag and skip subsequent local writes?
> 
> wdyt?

Agree that it's hard to follow, your understanding is correct. Of your two suggestions, I prefer the second (write local synchronously) because I don't want to increase load on the remote cache, so I'll make the changes here so that it behaves that way instead.

**Comment by iain-macdonald** [comment]
I've been trying to refactor this per earlier comments, but haven't found a good way to do so while keeping byte_stream_server_proxy a wrapper of byte_stream_server. IMO that's worth the added complexity because there's a bunch of stuff in byte_stream_server relating to error handling and compressed/uncompressed input/output that I'd rather not duplicate, and it also makes future changes to byte_stream_server pass on to the proxy server.

That said, I did refactor this to make the dual-streaming stuff a bit easier to follow (at least in my opinion). Curious what you think of the new version and my points about the composition? Happy to chat on Slack / huddle tomorrow for higher bandwidth too.

**Comment by iain-macdonald** [comment]
Alright, made the changes we discussed offline. This PR has ballooned a little bit, but most of it is moving code around / adding tests, so hopefully it's still a manageable review, but LMK if not.

**Comment by iain-macdonald** [comment]
Gonna break this up into a handful of smaller PRs, stay tuned.

File: enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go [hunk 1]
```diff
@@ -65,33 +148,67 @@ func (s *ByteStreamServerProxy) Read(req *bspb.ReadRequest, stream bspb.ByteStre
...
+		// Send to the local ByteStreamServer (if it hasn't errored)
+		localDone := req.GetFinishWrite()
+		if local != nil {
+			if err := local.Send(req); err != nil {
+				if err == io.EOF {
+					localDone = true
+				} else {
+					log.Infof("error writing to local bytestream server: %s", err)
+				}
+				local = nil
```
Comment (tylerwilliams) [inline enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go:175]: what's the reason for nilling out local here?

**Review by tylerwilliams — CHANGES_REQUESTED** [review]
this looks a lot better than before, thanks for making those changes!

File: enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go [hunk 2]
```diff
@@ -6,57 +6,140 @@ import (
...
+			// If some responses were streamed to the client, just return the
+			// error. Otherwise, fall-back to remote.
+			if responseSent {
+				return err
+			} else {
+				return s.readRemote(req, stream)
+			}
+		}
+
+		stream.Send(rsp)
```
[the remark this answers is not in the document]
Comment (iain-macdonald) [reply enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go]: whoops, yeah, fixed

File: enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go [hunk 3]
```diff
@@ -6,57 +6,140 @@ import (
...
+	if err != nil {
+		return err
+	}
+
+	var localWriteStream *localWriter = nil
+	if req.ReadOffset == 0 {
+		localStream, err := s.local.Write(stream.Context())
+		if err != nil {
+			log.Warningf("error writing to local bytestream server: %s", err)
+			localWriteStream = nil
```
[the remark this answers is not in the document]
Comment (iain-macdonald) [reply enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go]: nah, fixed

File: enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go [hunk 4]
```diff
@@ -6,57 +6,140 @@ import (
...
 		if err == io.EOF {
+			if localWriteStream != nil {
+				localWriteStream.close()
+			}
 			break
 		}
 		if err != nil {
 			return err
 		}
+
```
[the remark this answers is not in the document]
Comment (iain-macdonald) [reply enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go:140]: whoops, flipped this around so that stream is closed for any error

File: enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go (hunk 1 above)
Comment (iain-macdonald) [reply enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go:184]: to prevent sending addition frames to the local server if it returned an error

File: enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go [hunk 5]
```diff
@@ -65,33 +148,67 @@ func (s *ByteStreamServerProxy) Read(req *bspb.ReadRequest, stream bspb.ByteStre
...
 	for {
 		req, err := stream.Recv()
 		if err != nil {
 			return err
 		}
-		writeDone := req.GetFinishWrite()
-		if err := remoteStream.Send(req); err != nil {
+
+		// Send to the local ByteStreamServer (if it hasn't errored)
+		localDone := req.GetFinishWrite()
```
[the remark this answers is not in the document]
Comment (iain-macdonald) [reply enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go:176]: yeah, i think it's effectively intended to be an OR of req.GetFinishWrite() and local.Send() == io.EOF. Copied from here: https://github.com/buildbuddy-io/buildbuddy/blob/3eff2d5d76c811c0bdc2a659cdd76acc87db0fed/server/cache_proxy/cache_proxy.go#L262-L269

**Review by tylerwilliams — APPROVED** [review]
lgtm -- can you get vadim's approval on the streaming loop specifically before submitting?

File: enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go [hunk 6]
```diff
@@ -6,92 +6,218 @@ import (
...
+	return err
+}
+
+func (s *ByteStreamServerProxy) readRemote(req *bspb.ReadRequest, stream bspb.ByteStream_ReadServer) error {
+	remoteReadStream, err := s.remote.Read(stream.Context(), req)
+	if err != nil {
+		return err
+	}
+
+	var localWriteStream *localWriter = nil
```
Comment (vadimberezniker) [inline enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go:114]: nit: don't need ` = nil`. pointers are always nil initialized.

Comment (vadimberezniker) [inline enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go:114]: instead of initializing this to nil, would the code be cleaner if you instead instantiated a discardingLocalWriter{} that discards all data?

File: enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go [hunk 7]
```diff
@@ -6,92 +6,218 @@ import (
...
+	if req.ReadOffset == 0 {
+		localStream, err := s.local.Write(stream.Context())
+		if err != nil {
+			log.Warningf("error writing to local bytestream server: %s", err)
```
Comment (vadimberezniker) [inline enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go:118]: did you mean to initialize localWriteStream with a nil localStream? I think that would lead to a panic.

File: enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go [hunk 8]
```diff
@@ -6,92 +6,218 @@ import (
...
+			initialized:  false,
+			offset:       int64(0),
+		}
+	}
+
+	for {
+		rsp, err := remoteReadStream.Recv()
+		if err != nil {
+			if localWriteStream != nil {
+				localWriteStream.close()
```
Comment (vadimberezniker) [inline enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go:133]: should this function be called commit instead of close()? even though this is an error case, close() will attempt to finish the write. the write would fail, but it doesn't seem like we should be trying to commit it in the first place on error.

File: enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go [hunk 9]
```diff
@@ -6,92 +6,218 @@ import (
...
+		remote: remote,
 	}, nil
 }
 
 func (s *ByteStreamServerProxy) Read(req *bspb.ReadRequest, stream bspb.ByteStream_ReadServer) error {
-	remoteStream, err := s.remoteCache.Read(stream.Context(), req)
+	localReadStream, err := s.local.Read(stream.Context(), req)
 	if err != nil {
-		return err
+		return s.readRemote(req, stream)
```
Comment (vadimberezniker) [inline enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go:50]: I think you should be generous about logging on error-cases instead of silently ignoring them. Will make problems easier to debug in the future. Also please use Ctx* log functions where possible.

File: enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go [hunk 10]
```diff
@@ -6,92 +6,218 @@ import (
...
+			}
+			if err == io.EOF {
+				break
+			}
 			return err
 		}
+
+		if localWriteStream != nil {
+			if err := localWriteStream.send(rsp.Data); err != nil {
+				log.Debugf("Error writing locally: %s", err)
```
Comment (vadimberezniker) [inline enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go:143]: I think should at least be an Info log so that we see it in prod as well? Same below.

File: enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy_test.go [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+		// uploads)
+		tc.bazelVersion = "5.0.0"
+		t.Run(tc.name+", bazel "+tc.bazelVersion, run)
+
+		tc.bazelVersion = "5.1.0"
+		t.Run(tc.name+", bazel "+tc.bazelVersion, run)
+	}
+}
+
+func newUUID(t *testing.T) string {
```
Comment (vadimberezniker) [inline enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy_test.go:331]: nit: might as well use uuid.New() instead

**Review by vadimberezniker — CHANGES_REQUESTED** [review]

File: enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go (hunk 9 above)
Comment (iain-macdonald) [reply enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go:54]: sg, done

File: enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go (hunk 6 above)
Comment (iain-macdonald) [reply enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go]: whoops, thanks!

Comment (iain-macdonald) [reply enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go]: sure, done

File: enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go (hunk 7 above)
Comment (iain-macdonald) [reply enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go]: oh nice catch, fixed

File: enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go (hunk 8 above)
Comment (iain-macdonald) [reply enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go]: renamed and restructured to not call on errors except EOF

File: enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go (hunk 10 above)
Comment (iain-macdonald) [reply enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go]: sg, done

File: enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy_test.go (hunk 11 above)
Comment (iain-macdonald) [reply enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy_test.go]: sure, fixed

**Review by vadimberezniker — APPROVED** [review]
